kernel/src/actions/mod.rs

```diff
@@ -221,25 +227,25 @@ impl Protocol {
     }
 
     /// Get the reader features for the protocol
-    pub fn reader_features(&self) -> Option<&[String]> {
+    pub fn reader_features(&self) -> Option<&[ReaderFeatures]> {
         self.reader_features.as_deref()
     }
 
     /// Get the writer features for the protocol
-    pub fn writer_features(&self) -> Option<&[String]> {
+    pub fn writer_features(&self) -> Option<&[WriterFeatures]> {
         self.writer_features.as_deref()
     }
 
     /// True if this protocol has the requested reader feature
-    pub fn has_reader_feature(&self, feature: &ReaderFeatures) -> bool {
+    pub fn has_reader_feature(&self, feature: ReaderFeatures) -> bool {
         self.reader_features()
-            .is_some_and(|features| features.iter().any(|f| f == feature.as_ref()))
+            .is_some_and(|features| features.iter().any(|f| *f == feature))
     }
 
     /// True if this protocol has the requested writer feature
-    pub fn has_writer_feature(&self, feature: &WriterFeatures) -> bool {
+    pub fn has_writer_feature(&self, feature: WriterFeatures) -> bool {
         self.writer_features()
-            .is_some_and(|features| features.iter().any(|f| f == feature.as_ref()))
+            .is_some_and(|features| features.iter().any(|f| *f == feature))
     }
```

Comment:
these seem like reasonable (but orthogonal) changes? i think taking ownership for an enum which implements Copy is more idiomatic but not sure if we should do it all here?

Reply:
oh weird - well either way it feels like having an owned feature is a good change - just unsure if we want to include in this PR

Reply:
and honestly, passing a reference here seems off to me, It should just be an owned value and compared as such right?

Reply:
yep exactly i think owned is better since it's `Copy`. this is fine we can leave it in - just trying to mitigate mixing different changes in one PR too much bc that usually just slows things down

Reply:
Once we have a string in the variant, is it still `Copy` though?

Reply:
No it is not because the underlying fields are not copy.